recipes/LibriSpeech/ASR/transformer/train_bayesspeech.py

```diff
@@ -0,0 +1,565 @@
+                [tokens_eos_lens, tokens_eos_lens], dim=0
+            )
+            tokens = torch.cat([tokens, tokens], dim=0)
+            tokens_lens = torch.cat([tokens_lens, tokens_lens], dim=0)
+
+        loss_seq = self.hparams.seq_cost(
+            p_seq, tokens_eos, length=tokens_eos_lens
+        ).sum()
+
+        # now as training progresses we use real prediction from the prev step instead of teacher forcing
+
+        loss_ctc = self.hparams.ctc_cost(
+            p_ctc, tokens, wav_lens, tokens_lens
+        ).sum()
+
+        loss = (
+            self.hparams.ctc_weight * loss_ctc
+            + (1 - self.hparams.ctc_weight) * loss_seq
+            + self.hparams.kl_div_weight * self.modules.Transformer.kl_div
+        )
+
+        if stage != sb.Stage.TRAIN:
+            current_epoch = self.hparams.epoch_counter.current
+            valid_search_interval = self.hparams.valid_search_interval
+            if current_epoch % valid_search_interval == 0 or (
+                stage == sb.Stage.TEST
+            ):
+                # Decode token terms to words
+                predicted_words = [
+                    tokenizer.decode_ids(utt_seq).split(" ") for utt_seq in hyps
+                ]
+                target_words = [wrd.split(" ") for wrd in batch.wrd]
+                self.wer_metric.append(ids, predicted_words, target_words)
+
+            # compute the accuracy of the one-step-forward prediction
+            self.acc_metric.append(p_seq, tokens_eos, tokens_eos_lens)
+        return loss
+
+    def fit_batch(self, batch):
+
+        should_step = self.step % self.grad_accumulation_factor == 0
+        # Managing automatic mixed precision
+        if self.auto_mix_prec:
+            self.optimizer.zero_grad()
+            with torch.cuda.amp.autocast():
+                outputs = self.compute_forward(batch, sb.Stage.TRAIN)
+                loss = self.compute_objectives(outputs, batch, sb.Stage.TRAIN)
+            self.scaler.scale(loss / self.grad_accumulation_factor).backward()
+            if should_step:
+                self.scaler.unscale_(self.optimizer)
+                if self.check_gradients(loss):
+                    self.scaler.step(self.optimizer)
+                self.scaler.update()
+                self.optimizer_step += 1
+
+                # anneal lr every update
+                self.hparams.noam_annealing(self.optimizer)
+        else:
+            outputs = self.compute_forward(batch, sb.Stage.TRAIN)
+            loss = self.compute_objectives(outputs, batch, sb.Stage.TRAIN)
+            (loss / self.grad_accumulation_factor).backward()
+            if should_step:
+                if self.check_gradients(loss):
+                    self.optimizer.step()
+                self.optimizer.zero_grad()
+                self.optimizer_step += 1
+
+                # anneal lr every update
+                self.hparams.noam_annealing(self.optimizer)
+
+        return loss.detach().cpu()
```

Comment:
I'm curious, is the only reason for reproducing the `fit_batch()` here in order to add the `noam_annealing()`? This could better be done with the new function `on_fit_batch_end()`. If this code was inherited from another recipe, that one should eventually get updated too.

Reply:
It was indeed copied from the `train.py` script in the same directory. I guess the reason is that we want to anneal the learning rate at each step, but only if the `should_step` flag is `True`.

Reply:
In theory it should help neither with training time nor convergence speed, since wall-clock time for each epoch is definitely larger (2x parameters for each Bayesian layer, requires sampling for each forward pass, etc.) and we are over-regularizing (LibriSpeech has plenty of data while Bayesian neural networks are good in low data regimes).

Reply:
I trained both Bayesian transformer and standard one for 20 epochs with same batch size, this is the log:
[bayesspeech_vs_transformer.txt](https://github.com/speechbrain/speechbrain/files/10974121/bayesspeech_vs_transformer.txt)

Reply:
That's a neat chart, looks like Bayes Speech has better validation accuracy until ~epoch 7 or 8. I wonder how this might work alongside hyperparameter search or neural architecture search to reduce computational budget... Food for thought anyway!